File: project/mobile/lib/services/tracking_controller.dart
// Inti logika tracking. Server = source of truth.
// - STANDBY: GPS OFF, polling ringan (default 60 dtk).
// - TRACKING: GPS tiap interval server (default 30 dtk), buffer -> batch sync.
// - Offline: point masuk SQLite queue, sync saat internet kembali.
// - Tidak ada tombol start/stop: semua mengikuti status server.
import 'dart:async';
import 'dart:io';
import 'package:battery_plus/battery_plus.dart';
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:geolocator/geolocator.dart';
import 'package:uuid/uuid.dart';

import 'api_service.dart';
import 'offline_queue.dart';
import 'storage_service.dart';

class TrackingController {
  TrackingController._();
  static final TrackingController instance = TrackingController._();

  final _battery = Battery();
  bool _running = false;
  bool _tracking = false;
  DateTime _lastPoll = DateTime.fromMillisecondsSinceEpoch(0);
  DateTime _lastGps = DateTime.fromMillisecondsSinceEpoch(0);
  int _pollInterval = 60;
  int _gpsInterval = 30;
  bool _wasTracking = false;
  bool _netOnline = true;

  static const int batchSize = 5;

  Future<void> start() async {
    if (_running) return;
    _running = true;
    Connectivity().onConnectivityChanged.listen((results) {
      final online = results.any((r) => r != ConnectivityResult.none);
      if (online != _netOnline) {
        _netOnline = online;
        _sendEventSafe(online ? 'NETWORK_ONLINE' : 'NETWORK_OFFLINE');
        if (online) _syncQueue();
      }
    });
    Timer.periodic(const Duration(seconds: 10), (_) => tick());
    await _sendEventSafe('APP_STARTED');
    await tick();
  }

  Future<void> tick() async {
    final storage = StorageService.instance;
    final token = await storage.deviceToken;
    if (token == null) return; // belum diapprove / belum aktivasi

    final now = DateTime.now();
    if (now.difference(_lastPoll).inSeconds < _pollInterval && !_tracking) {
      return;
    }
    if (_tracking && now.difference(_lastGps).inSeconds < _gpsInterval) {
      return;
    }

    int? batt;
    try { batt = await _battery.batteryLevel; } catch (_) {}

    // --- Poll status server ---
    try {
      final status = await ApiService.instance.statusByToken(token, battery: batt);
      _lastPoll = now;
      _pollInterval = status.standbyPollInterval;
      _gpsInterval = status.trackingInterval;

      if (status.trackingRequired && !_wasTracking) {
        _wasTracking = true;
        _sendEventSafe('TRACKING_STARTED');
      } else if (!status.trackingRequired && _wasTracking) {
        _wasTracking = false;
        _sendEventSafe('TRACKING_STOPPED');
      }
      _tracking = status.trackingRequired;

      await storage.writeUiState({
        'state': _tracking ? 'TRACKING' : 'STANDBY',
        'session_names': status.activeSessions.map((s) => s.name).join(', '),
        'server_time': status.serverTime,
        'battery': '${batt ?? ''}',
      });
    } on ApiException catch (e) {
      if (e.statusCode == 401 || e.statusCode == 403) {
        // Device revoked / token tidak valid: berhenti mengirim GPS.
        _tracking = false;
        await storage.writeUiState({'state': 'REVOKED'});
        return;
      }
      _tracking = _wasTracking; // server error: pertahankan state terakhir
    } catch (_) {
      // Tidak ada internet: GPS tetap jalan jika tracking, data masuk queue.
      await StorageService.instance.writeUiState({'net_ok': 'OFF'});
    }

    // --- Ambil GPS saat tracking ---
    if (_tracking && now.difference(_lastGps).inSeconds >= _gpsInterval) {
      await _capturePoint(batt);
      _lastGps = now;
    }

    // --- Batch sync ---
    if (_netOnline) {
      await _syncQueue();
    }
  }

  Future<void> _capturePoint(int? batt) async {
    try {
      final serviceOn = await Geolocator.isLocationServiceEnabled();
      if (!serviceOn) {
        _sendEventSafe('GPS_DISABLED', battery: batt);
        await StorageService.instance.writeUiState({'gps_ok': 'OFF'});
        return;
      }
      final pos = await Geolocator.getCurrentPosition(
        locationSettings: const LocationSettings(accuracy: LocationAccuracy.high),
      );
      await OfflineQueue.instance.enqueue({
        'client_point_id': const Uuid().v4(),
        'latitude': pos.latitude,
        'longitude': pos.longitude,
        'accuracy': pos.accuracy,
        'altitude': pos.altitude,
        'speed': pos.speed,
        'battery': batt,
        'recorded_at': _fmt(pos.timestamp),
      });
      await StorageService.instance.writeUiState({
        'gps_ok': 'ON',
        'last_update': _fmt(DateTime.now()),
      });
    } catch (e) {
      await StorageService.instance.writeUiState({'gps_ok': 'OFF'});
    }
  }

  Future<void> _syncQueue() async {
    final token = await StorageService.instance.deviceToken;
    if (token == null) return;
    final pending = await OfflineQueue.instance.pending(limit: 50);
    if (pending.isEmpty) return;

    // Kirim batch 5-10 point per request
    final chunks = <List<Map<String, dynamic>>>[];
    for (var i = 0; i < pending.length; i += batchSize) {
      chunks.add(pending.sublist(i, i + batchSize > pending.length ? pending.length : i + batchSize));
    }
    for (final chunk in chunks) {
      try {
        await ApiService.instance.syncLocations(token, chunk.map((p) => {
          'client_point_id': p['client_point_id'],
          'latitude': p['latitude'],
          'longitude': p['longitude'],
          'accuracy': p['accuracy'],
          'altitude': p['altitude'],
          'speed': p['speed'],
          'battery': p['battery'],
          'recorded_at': p['recorded_at'],
        }).toList());
        await OfflineQueue.instance.removeIds(chunk.map((p) => p['id'] as int).toList());
        await StorageService.instance.writeUiState({'net_ok': 'ON'});
      } catch (_) {
        break; // internet mati lagi / server error: coba di tick berikutnya
      }
    }
  }

  Future<void> _sendEventSafe(String type, {int? battery}) async {
    try {
      final token = await StorageService.instance.deviceToken;
      if (token == null) return;
      await ApiService.instance.sendEvent(token, type, battery: battery);
    } catch (_) {}
  }

  static Future<Map<String, String>> deviceInfo() async {
    final info = DeviceInfoPlugin();
    if (Platform.isAndroid) {
      final a = await info.androidInfo;
      return {'platform': 'android', 'model': '${a.manufacturer} ${a.model}'};
    }
    if (Platform.isIOS) {
      final i = await info.iosInfo;
      return {'platform': 'ios', 'model': i.utsname.machine};
    }
    return {'platform': 'unknown', 'model': 'unknown'};
  }

  static String _fmt(DateTime t) =>
      '${t.year.toString().padLeft(4, '0')}-${t.month.toString().padLeft(2, '0')}-${t.day.toString().padLeft(2, '0')} '
      '${t.hour.toString().padLeft(2, '0')}:${t.minute.toString().padLeft(2, '0')}:${t.second.toString().padLeft(2, '0')}';
}

File: project/mobile-rn/src/tracking/trackingController.ts
// Inti logika tracking. Server = source of truth.
// STANDBY: GPS OFF, polling ringan (~60 dtk, dari server).
// TRACKING: GPS tiap interval server (~30 dtk), buffer -> batch sync 5-10 titik.
// Offline: titik masuk SQLite queue; sync saat internet kembali.
// Revoked (401/403): hentikan tracking + hapus token lokal.
import NetInfo from '@react-native-community/netinfo';
import Geolocation from 'react-native-geolocation-service';
import BackgroundService from 'react-native-background-actions';
import {api} from '../api/client';
import {ApiError, LocationPoint} from '../types';
import {getDeviceToken, clearDeviceToken} from '../storage/secure';
import {enqueue, pending, removeIds, queueCount} from '../storage/queue';
import {getBatteryLevel} from '../device/deviceInfo';
import {sendDeviceEvent} from '../services/events';
import {setUiState} from '../services/uiState';
import {formatServerTime, newClientPointId} from '../utils/time';

const BATCH_SIZE = 8;

let running = false;
let tracking = false;
let pollIntervalSec = 60;
let gpsIntervalSec = 30;
let lastPoll = 0;
let lastGps = 0;
let netOnline = true;
let lastBatteryEventSent = 0;

function nowSec(): number {
  return Math.floor(Date.now() / 1000);
}

export function initTracking(): void {
  if (running) {
    return;
  }
  running = true;

  NetInfo.addEventListener(state => {
    const online = !!state.isConnected;
    if (online !== netOnline) {
      netOnline = online;
      setUiState({netOk: online});
      sendDeviceEvent(online ? 'NETWORK_ONLINE' : 'NETWORK_OFFLINE');
      if (online) {
        void syncQueue();
      }
    }
  });

  sendDeviceEvent('APP_STARTED');
  // Foreground tick: dilewati jika background service sedang berjalan (hindari GPS ganda).
  setInterval(() => {
    if (!BackgroundService.isRunning()) {
      void trackingTick();
    }
  }, 10000);
  void trackingTick();
}

export async function trackingTick(): Promise<void> {
  const token = await getDeviceToken();
  if (!token) {
    return; // belum aktivasi / belum diapprove
  }
  const now = nowSec();
  if (!tracking && now - lastPoll < pollIntervalSec) {
    return;
  }
  if (tracking && now - lastGps < gpsIntervalSec) {
    return;
  }

  const battery = await getBatteryLevel();

  try {
    const status = await api.statusByToken(token, battery);
    lastPoll = now;
    pollIntervalSec = status.standby_poll_interval || 60;
    gpsIntervalSec = status.tracking_interval || 30;

    if (status.tracking_required && !tracking) {
      tracking = true;
      await sendDeviceEvent('TRACKING_STARTED');
    } else if (!status.tracking_required && tracking) {
      tracking = false;
      await sendDeviceEvent('TRACKING_STOPPED');
    }
    tracking = status.tracking_required;

    setUiState({
      state: tracking ? 'TRACKING' : 'STANDBY',
      personnelName: status.personnel?.name ?? '',
      nrp: status.personnel?.nrp ?? '',
      rank: status.personnel?.rank ?? '',
      sessions: status.active_sessions.map(s => s.name),
      battery: battery ?? null,
      serverTime: status.server_time,
      netOk: true,
    });

    // Battery event (threshold crossing saja, throttle 30 mnt)
    if (battery !== undefined && battery <= 15 && now - lastBatteryEventSent > 1800) {
      lastBatteryEventSent = now;
      await sendDeviceEvent('BATTERY_LOW');
    }
  } catch (e) {
    const err = e as ApiError;
    if (err.status === 401 || err.status === 403) {
      // Token dicabut / device revoked -> stop total.
      tracking = false;
      await clearDeviceToken();
      setUiState({state: 'REVOKED'});
      return;
    }
    setUiState({netOk: false}); // offline: GPS tetap jalan jika tracking
  }

  if (tracking && now - lastGps >= gpsIntervalSec) {
    await capturePoint(battery);
    lastGps = now;
  }
  if (netOnline) {
    await syncQueue();
  }
}

function capturePoint(battery?: number): Promise<void> {
  return new Promise(resolve => {
    Geolocation.getCurrentPosition(
      async pos => {
        const point: LocationPoint = {
          client_point_id: newClientPointId(),
          latitude: pos.coords.latitude,
          longitude: pos.coords.longitude,
          accuracy: pos.coords.accuracy ?? null,
          altitude: pos.coords.altitude ?? null,
          speed: pos.coords.speed ?? null,
          battery: battery ?? null,
          recorded_at: formatServerTime(new Date(pos.timestamp)),
        };
        await enqueue(point);
        setUiState({
          gpsOk: true,
          lastSync: formatServerTime(new Date()),
          queueCount: await queueCount(),
        });
        resolve();
      },
      async err => {
        setUiState({gpsOk: false});
        if (err.code === 2) {
          // 2 = POSITION_UNAVAILABLE
          await sendDeviceEvent('GPS_DISABLED');
        }
        resolve();
      },
      {enableHighAccuracy: true, timeout: 15000, maximumAge: 5000},
    );
  });
}

export async function syncQueue(): Promise<void> {
  const token = await getDeviceToken();
  if (!token) {
    return;
  }
  const rows = await pending(50);
  if (!rows.length) {
    setUiState({queueCount: 0});
    return;
  }
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const chunk = rows.slice(i, i + BATCH_SIZE);
    try {
      await api.syncLocations(
        token,
        chunk.map(({id, ...p}) => p),
      );
      // Hapus lokal HANYA setelah server konfirmasi berhasil.
      await removeIds(chunk.map(c => c.id));
    } catch {
      break; // gagal -> titik tetap di queue, coba tick berikutnya
    }
  }
  setUiState({queueCount: await queueCount()});
}
